### ISSUE TITLE
FR: Android TV

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#5725]
I know, I know. It is as stupid as it sounds. Who wants to run an e-reader app on a tv?.

Technically is possible and pretty easy: just detect if we run on a tv platform, set hasDpad=true and change orientation from "sensorPortrait" to "landscape".

I'm not sure if there's some hardcoded value somewhere (on the filemanager?) for height > width. Today tried to run an almost "vanilla" version on my new tv box and while the splash screen was in landscape mode once it reached the filemanager it turned to reverse portrait mode.

**Comment by Frenzie** [comment]
Depends a bit on the kind of document. ;-)

The FileManager goes through some trouble to ensure portrait iirc.

**Comment by pazos** [comment]
> The FileManager goes through some trouble to ensure portrait iirc.

Awh. I was expecting some kind of magic port without any effort  :smile:

**Comment by NiLuJe** [comment]
Yeah, IIRC, the FM is very explicitly Portrait only. It can do Inverted Portrait (but you can only achieve that on devices with a Gyro, f.g., Forma/Libra).

**Comment by KenMaltby** [comment]
I wonder about the latest game machines, like the 4K Xbox One, they seem to be headed into a more general app support approach, than just game machines.  Not as obvious as Android, and there could be Microsoft licensing issues, but might be a future thing, especially combined with the voice activated gadgets people are letting into their lives.
Luck;
Ken

**Comment by NiLuJe** [comment]
Unfortunately, those platforms are extremely hostile to homebrew. You basically have to be a registered developer, access to the SDKs usually costs big money, and they're an NDA riddled mess that makes everything harder (impossible?N) for OSS projects...

**Comment by Frenzie** [comment]
Yeah, the Xbox One could be seen as a locked down nightmare version of Windows. They'd make for a pretty interesting PC otherwise, as a viable alternative for one of those € 300 or possibly even as much as € 400 NUCs. Except they only cost a little over € 200.

Unfortunately it's not just the software, but also the hardware that's purposefully distinct in minute details.

**Comment by pazos** [comment]
Reading PDFs on a big tv:

![test](https://user-images.githubusercontent.com/975883/71689985-af2b6f80-2da3-11ea-955d-0d284e4ce7f9.gif)

The shadows you see here are not visible in the tv. I guess is something related to scrcpy on androidTv.
The dpad controls are rotated.

**Comment by Frenzie** [comment]
Much better on real Linux with an Xbox One or Steam controller. Or KDE Connect. :-P

![2020-01-02 21 18 06](https://user-images.githubusercontent.com/202757/71690644-92903700-2da5-11ea-8e5c-9ea03af22f31.jpg)

**Comment by Frenzie** [comment]
> I'm unable to reach the "home" and "plus" buttons on the FM. Not really aware of how its done on devices without touchscreen.

You have a special plus menu in the top menu instead. But yeah, that's only enabled by default on ancient non-touch Kindles I believe.

**Comment by Frenzie** [comment]
You see it if you switch your SDL device to isTouchDevice = no.

![Screenshot_2020-01-02_21-22-28](https://user-images.githubusercontent.com/202757/71690802-06cada80-2da6-11ea-8614-50bb0716f5a4.png)

**Comment by pazos** [comment]
Ok, that makes sense. I forgot about isTouchDevice.

**Comment by rokejulianlockhart** [comment]
But what if the television in mounted vertically? It might sound insane because it is, but VESA mounts make it remarkably easy.

**Comment by axu2** [comment]
Would be pretty cool, for now using the Pocketbook TV reader app. It works decent.

**Comment by pazos** [comment]
@xmha97 @rokejulianlockhart @axu2 

The apk already works in AndroidTV (for a couple of years, at least).

Everything works, including rotation so vertical panels are not a problem (as long as you setup it correctly, it should start on landscape cause it's the native orientation on tvs). 

The only reason this ticket is still open is because the app cannot be fully used with the controls on the remote. See my suggestion in https://github.com/koreader/koreader/pull/6327#discussion_r450787837.

**Comment by pazos** [comment]
I would love a helping hand here, @Commodore64user

A bit of context: xupport for android TV is done except for user input, which is quite limited using the tv controller.
Nowadays using even the 10% of KO features would require a mouse, keyboard or gamepad.

I see you're doing a lot of Non Touch work. You can consider this a NT expert-seeking question :).

Are you able to figure out a way to do basic KO navigation with a 5-way d-pad and a hardcoded **back** button?

I'm not asking for the implementation. Just the idea :) . Yeah, I know that's a lot harder.

**Comment by Commodore64user** [comment]
> Are you able to figure out a way to do basic KO navigation with a 5-way d-pad and a hardcoded back button?

I think it's possible although Father Christmas won't like it as you'd have to be a naughty boy and take an unconventional route. In all seriousness though, the only difficult place I see is the menus really (top and bottom) where ideally I would want to have one extra button for long-press actions... I guess you could have something like

### menu

**back**: back
**up**: up
**down**: down
**left**: left
**right**: long-press (we need to sacrifice here)
**centre button**: select or press

file manager only uses left/right for navigation in `Mosaic` mode, but it is still usable with up-down only so...

### file manager

**back**: back
**up**: up
**down**: down
**left**: top menu
**right**: long-press
**centre button**: select or press

remember to flag AndroidTV as `not isTouchDevice` and you'll get the plus menu in the top menu.
and surprisingly there are a lot of options here

### reader

instead of having two buttons for top and bottom menu, I would go with one button triggering a hardcoded `QuickMenu` with both top and bottom menus there (and perhaps other actions you might want quick access to) or you could have direct access to the top menu and add the bottom menu as an extra icon like the plus menu in FM. I personally thing the first option is better... (feel free to swap up/down with left/right for consistency)

**back**: back
**up**: home (this way you can use back to move to saved location and the like)
**down**: quick menu (top and bottom menu)
**left**: previous page
**right**: next page
**centre button**: I would map this to the `startHighlightIndicator` event, i.e highlights, dictionary lookups all that good stuff.

that should get you almost everything you'd want, and if there is anything else, like bookmarking, you could always add it to the `QuickMenu`

**Comment by itsjakobx** [comment]
Some Family members of mine are also suddenly interested in reading books on their TV. I would love to use KOReader for it as it has opds that i can connect to my audiobookshelf.
